# --------------------------------------- Workshop #2 -------------------------------------


* This workshop includes marked tasks that comprise 16% of your final mark in this module. 

* To complete the tasks, you need to apply data preprocessing methods discussed in Lecture 2. However, you may need to research to find solutions to some tasks. 

* You can duplicate the code and report cells if you need more than one code/report cell for your solution

# Tasks

## TASK 2.1: Download the adult dataset from Canvas, import it into Jupyter Notebook, and complete the following steps (Completing the report cell is required):

### a. Write a code to show how many Null values each column contains. Report the columns that contain Null values in the report cell (Hint: use the isna().sum() method to show the number of the Null values) (NOTE: Nan values are also considered as Null values) (1%). 

### b. Change the display setting and print the first 100 rows. Find columns that contain wrong data and report them in the report cell (1%). 


### c. We cannot inspect the entire data by eyes for columns containing wrong data. Explain in the report cell what method we could use to find all columns which contain wrong data (2%)


2.1

In [18]:

import pandas as pd

file_path = '/Users/maz/7CS033DataMining/Week2/Workshop/adult.csv'

adult_data = pd.read_csv(file_path)

adult_data.head()

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K


**<u>2.1 a)</u>** 

<span style="font-size: 16px; white-space-collapse: preserve;">In pandas, you may use the isna() and sum() methods in order to check for null (or NaN) values in each column of the dataset and then report the columns. One can determine which columns contain null values by using the count of null values that the function returns. In order to accomplish this, we will use the pandas library in Python code to generate a report that specifies the locations of all null values in each column of our dataset as well as a total for all null values in each column. After a short report detailing which columns have null values, data cleaning and preparation can start.</span>

In [19]:
null_values = adult_data.isna().sum()

print(null_values)

age                 0
workclass         963
fnlwgt              0
education           0
education-num       0
marital-status      0
occupation        966
relationship        0
race                0
sex                 0
capital-gain        0
capital-loss        0
hours-per-week      0
native-country    274
income              0
dtype: int64


This code will locate and calculate each column in your dataset with any null (NaN) value when it is executed in the Python environment.

We ran the aforementioned code for each of the following dataset columns and discovered null values there:

 workclass: The type of employment for the individuals is represented by 963 null values with missing values. 

occupation: There are 966 null values in occupation with missing values pertaining to the job title information. 

native-country: There are 274 null values pertaining to missing data regarding some persons' native country.  

**Conclusions and Next Steps**

When handling null values in those columns—workclass, occupation, and native country—during the data cleaning and preparation stage, some astute observation is needed. The following are some options for dealing with these absent entries: 

Such nulls are imputed using the mean or most frequent value (when numerical data are present) or with a clear "unknown" filler. 

We eliminated rows containing null values in which the missing data only made up a minor portion of the dataset and removing them did not materially alter the dataset. 

The null values must be handled before moving on to any other data analysis or machine learning tasks. <span style="font-family: -apple-system, BlinkMacSystemFont, sans-serif; color: var(--vscode-foreground);">Only then can the validity and integrity of any conclusions or models drawn from this dataset be guaranteed.</span>

**<u>2.1 b)</u>**

<span style="font-size: 16px; white-space-collapse: preserve;">For example, when printing the DataFrame or using the.head() method, you will need to modify the pandas display settings to allow for more visible rows in order to print the first 100 rows of the dataset. By default, pandas use a truncated view for presentation, which saves space by hiding certain rows in the centre of the DataFrame. You can modify this by calling pd.set_option().</span>

In [20]:
import pandas as pd

file_path = '/Users/maz/7CS033DataMining/Week2/Workshop/adult.csv'
adult_data = pd.read_csv(file_path)

pd.set_option('display.max_rows', 100)

print(adult_data.head(100))

    age         workclass  fnlwgt     education  education-num  \
0    39         State-gov   77516     Bachelors             13   
1    50  Self-emp-not-inc   83311     Bachelors             13   
2    38           Private  215646       HS-grad              9   
3    53           Private  234721          11th              7   
4    28           Private  338409     Bachelors             13   
5    37           Private  284582       Masters             14   
6    49           Private  160187           9th              5   
7    52  Self-emp-not-inc  209642       HS-grad              9   
8    31           Private   45781       Masters             14   
9    42           Private  159449     Bachelors             13   
10   37           Private  280464  Some-college             10   
11   30         State-gov  141297     Bachelors             13   
12   23           Private  122272     Bachelors             13   
13   32           Private  205019    Assoc-acdm             12   
14   40   

### <span style="font-size: 14px;"><b>Inspection for Wrong Data&nbsp;</b></span> 

Examination for Inaccurate Information After reviewing the 100 initial rows, you would check for any incorrect data. As an illustration: 

**Task 2.1b Report:** 

Checking for Inaccurate Information We have discovered possible problems in the columns listed below after going over the dataset's first 100 entries: 

**Column\_Name\_1**: Had problems with inconsistent data format or undocumented placeholders. 

**Column\_Name\_2**: Similarly, any indications of inaccurate data or placeholders that point to information that is missing. 

**Advice**: As a systematic approach across the whole dataset, an automated method such as value\_counts() for categorical data to see if there are any unusual categories and summary statistics for numerical data to identify any outliers or invalid entries is suggested. Additionally helpful are regular expressions and bespoke searches for known placeholders or improper formats.

Replace the descriptions for Column\_Name\_1, Column\_Name\_2, and the corresponding columns with your actual observations and results for the first 100 rows. This reporting phase is crucial because it provides a clear foundation for any further preprocessing processes you may take with the data and documents the cleaning and investigative process you have already completed.

**<u>2.1 c)</u>**

<span style="font-size: 16px; white-space-collapse: preserve;">In a dataset such as the, you would normally take a multi-step strategy that makes use of both automated approaches and domain knowledge to systematically detect the incorrect data from the columns. The term "inappropriate data" can refer to a wide range of items, including numbers that fall outside of an intended range and erroneous data types like strings in a numerical column or reserved words like "?" and "N/A" that pandas does not recognise as null.</span>

### Techniques for locating incorrect data in datasets usually combine automated and human inspection methods according to the unique properties of the data. These techniques may consist of:

1. **Data profiling**: is the process of looking for abnormalities or outliers in data distributions and summary statistics.
2. **Data Type Validation**: Verifying that each column's values correspond to the intended data types (e.g., making sure numerical values aren't accidentally entered as strings).
3. **Range checking**: Confirming that values are inside a permissible range that has been specified by the data specifications or domain expertise.
4. **Pattern recognition**: is the process of using regular expressions, particularly with text fields, to find values that do not fit a predetermined format.
5. **Comparing values**: from different columns or records to spot discrepancies, including dates that deviate from an event's expected timeline, is known as a consistency check.
6. **Duplicate detection**: involves finding and examining duplicate records that may point to incorrect data entry.
7. **Cross-referencing External Data**: Verifying information by comparing it to reliable outside sources to guarantee accuracy.
8. **Manual Review**: In certain situations, a subject matter expert's manual review may be required to find subtle problems that automated techniques are unable to detect.

Combining these techniques can assist in methodically locating and fixing inaccurate data in a dataset.

<span style="font-size: 16px; white-space-collapse: preserve;"><b>Summary Statistics: </b>The describe() method can be used to create summary statistics for numerical columns to start. In this manner, any abnormalities, such as negative results where only positive values are anticipated or abnormally high or low readings, will be easy to spot.</span>

In [3]:
adult_data.describe()

,age,fnlwgt,education-num,capital-gain,capital-loss,hours-per-week
count,48842.000000,4.884200e+04,48842.000000,48842.000000,48842.000000,48842.000000
mean,38.643585,1.896641e+05,10.078089,1079.067626,87.502314,40.422382
std,13.710510,1.056040e+05,2.570973,7452.019058,403.004552,12.391444
min,17.000000,1.228500e+04,1.000000,0.000000,0.000000,1.000000
25%,28.000000,1.175505e+05,9.000000,0.000000,0.000000,40.000000
50%,37.000000,1.781445e+05,10.000000,0.000000,0.000000,40.000000
75%,48.000000,2.376420e+05,12.000000,0.000000,0.000000,45.000000
max,90.000000,1.490400e+06,16.000000,99999.000000,4356.000000,99.000000


<span style="font-size: 16px; white-space-collapse: preserve;"><b>Value Counts: </b>To get a sense of the distribution and identify any clearly odd categories that may exist, apply the value_counts() method to each column of the categorical data. Make sure to find possible stand-ins for any missing or odd-looking data.</span>

In [5]:
adult_data['workclass'].value_counts()

workclass
Private             33906
Self-emp-not-inc     3862
Local-gov            3136
State-gov            1981
?                    1836
Self-emp-inc         1695
Federal-gov          1432
Without-pay            21
Never-worked           10
Name: count, dtype: int64

<span style="font-size: 16px; white-space-collapse: preserve;"><b>Data Types: </b>Make sure the data type in each column is correct by using dtypes. In this manner, the data type of each column may be guaranteed to be as intended (for example, numeric columns shouldn't have an object type because they contain strings).</span>

In [6]:
adult_data.dtypes

age                int64
workclass         object
fnlwgt             int64
education         object
education-num      int64
marital-status    object
occupation        object
relationship      object
race              object
sex               object
capital-gain       int64
capital-loss       int64
hours-per-week     int64
native-country    object
income            object
dtype: object

<span style="font-size: 16px; white-space-collapse: preserve;"><b>Custom Searches: </b>Use boolean indexing to filter the dataframe if you can identify certain placeholders or if you're working with known issues. For instance, if we are aware that in any categorical column, "?" denotes missing data, we might quickly count them using the following method:</span>

In [8]:
(adult_data['workclass'] == '?').sum()


1836

1. <span style="border: 0px solid rgb(217, 217, 227); box-sizing: border-box; --tw-border-spacing-x: 0; --tw-border-spacing-y: 0; --tw-translate-x: 0; --tw-translate-y: 0; --tw-rotate: 0; --tw-skew-x: 0; --tw-skew-y: 0; --tw-scale-x: 1; --tw-scale-y: 1; --tw-pan-x: ; --tw-pan-y: ; --tw-pinch-zoom: ; --tw-scroll-snap-strictness: proximity; --tw-gradient-from-position: ; --tw-gradient-via-position: ; --tw-gradient-to-position: ; --tw-ordinal: ; --tw-slashed-zero: ; --tw-numeric-figure: ; --tw-numeric-spacing: ; --tw-numeric-fraction: ; --tw-ring-inset: ; --tw-ring-offset-width: 0px; --tw-ring-offset-color: #fff; --tw-ring-color: rgba(69,89,164,.5); --tw-ring-offset-shadow: 0 0 transparent; --tw-ring-shadow: 0 0 transparent; --tw-shadow: 0 0 transparent; --tw-shadow-colored: 0 0 transparent; --tw-blur: ; --tw-brightness: ; --tw-contrast: ; --tw-grayscale: ; --tw-hue-rotate: ; --tw-invert: ; --tw-saturate: ; --tw-sepia: ; --tw-drop-shadow: ; --tw-backdrop-blur: ; --tw-backdrop-brightness: ; --tw-backdrop-contrast: ; --tw-backdrop-grayscale: ; --tw-backdrop-hue-rotate: ; --tw-backdrop-invert: ; --tw-backdrop-opacity: ; --tw-backdrop-saturate: ; --tw-backdrop-sepia: ; font-weight: 600; color: var(--tw-prose-bold);">Regular Expressions</span>: Use regular expressions to find patterns that do not match expected formats, especially useful for strings or dates.
    
2. <span style="border: 0px solid rgb(217, 217, 227); box-sizing: border-box; --tw-border-spacing-x: 0; --tw-border-spacing-y: 0; --tw-translate-x: 0; --tw-translate-y: 0; --tw-rotate: 0; --tw-skew-x: 0; --tw-skew-y: 0; --tw-scale-x: 1; --tw-scale-y: 1; --tw-pan-x: ; --tw-pan-y: ; --tw-pinch-zoom: ; --tw-scroll-snap-strictness: proximity; --tw-gradient-from-position: ; --tw-gradient-via-position: ; --tw-gradient-to-position: ; --tw-ordinal: ; --tw-slashed-zero: ; --tw-numeric-figure: ; --tw-numeric-spacing: ; --tw-numeric-fraction: ; --tw-ring-inset: ; --tw-ring-offset-width: 0px; --tw-ring-offset-color: #fff; --tw-ring-color: rgba(69,89,164,.5); --tw-ring-offset-shadow: 0 0 transparent; --tw-ring-shadow: 0 0 transparent; --tw-shadow: 0 0 transparent; --tw-shadow-colored: 0 0 transparent; --tw-blur: ; --tw-brightness: ; --tw-contrast: ; --tw-grayscale: ; --tw-hue-rotate: ; --tw-invert: ; --tw-saturate: ; --tw-sepia: ; --tw-drop-shadow: ; --tw-backdrop-blur: ; --tw-backdrop-brightness: ; --tw-backdrop-contrast: ; --tw-backdrop-grayscale: ; --tw-backdrop-hue-rotate: ; --tw-backdrop-invert: ; --tw-backdrop-opacity: ; --tw-backdrop-saturate: ; --tw-backdrop-sepia: ; font-weight: 600; color: var(--tw-prose-bold);">Visual Inspection</span>: While not feasible for large datasets, visualizing data through histograms, box plots, or scatter plots can help identify outliers or data inconsistencies.

### <span style="font-size: 16.38px;"><b><u>Reporting on Wrong Data Identification Method the methodology as follows:&nbsp;</u></b></span> 

### <span style="font-size: 16.38px;"><b><u>Identifying Wrong Data in the Dataset</u></b></span>

### <span style="font-size: 16.38px;">We must find and fix inaccurate data in our dataset in order to preserve the integrity of the analysis. This work can be completed using a variety of automated and manual techniques that address distinct types of data problems:&nbsp;</span> 

- <span style="font-size: 16.38px;"><b>Summary Statistics and Data Types:&nbsp;</b>An exploratory tool that is used to evaluate the data type and examine summary statistics initially aids in spotting errors in data types and anomalies in numerical data.&nbsp;</span>  
- <span style="font-size: 16.38px;"><b>Value Counts:&nbsp;</b>By looking at the distribution of categorical data, value counts can be used to determine whether any odd categories or placeholders for missing data exist.&nbsp;&nbsp;</span>    
- **Custom Searches and Regular Expressions:** <span style="font-size: 16.38px;">Are designed to allow users to look up certain patterns in the data that are there but not in the expected way, or to make searches after they become aware of a problem.&nbsp;</span>   
- <span style="font-size: 16.38px;"><b>Visual Inspection:&nbsp;</b></span> has the ability to highlight anomalies or discrepancies that are not obvious by numerical approaches, even if this will be an extremely difficult process for enormous data. Each phase of this process will be planned to jointly detect flaws, ranging from obvious outliers to minute discrepancies, in order to guarantee and attest to the accuracy and dependability of our dataset and serve as a foundation for additional investigation. 

  
<span style="color: var(--vscode-foreground); font-family: -apple-system, BlinkMacSystemFont, sans-serif;">This combines domain-specific expertise with automated data inspection techniques to fully detect errors in the data. Modify the details to reflect the adult's quirks and what you find in your dataset.The csv dataset is being examined.</span>

## TASK 2.2: Complete the following tasks (Completing the report cell is required):

### a. Determine whether the columns which contain Null values are numerical, nominal, or ordinal variables and report it in the report cell (1%).

### b. Delete all rows which contain Null values (1%). Then print the number of Null values of all columns using the isna().sum() method. 

### c. Import the dataset again. Find the mode of the categorical (i.e. nominal or ordinal) columns which contain Null values and use it to fill their Null values (2%)



<span style="color: rgb(55, 65, 81); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;"><b><u>2.2 a)</u></b></span>

<span style="font-size: 16px; white-space-collapse: preserve;">Let's first highlight the columns that include null entries and then discuss the nature of these columns based on a basic understanding of the dataset in order to better identify these columns and justify their categorization in numerical, nominal, or ordinal variables. Determine which set of columns (as previously shown) include null values, and then categorise each set based on the type of data that is typically found in each set.</span>  

<span style="font-size: 16px; white-space-collapse: preserve;">Step 1: Identify Columns with Null Value</span>

In [21]:
import pandas as pd

file_path = '/Users/maz/7CS033DataMining/Week2/Workshop/adult.csv'

adult_data = pd.read_csv(file_path)

null_values = adult_data.isna().sum()

columns_with_nulls = null_values[null_values > 0].index.tolist()

print("Columns with Null Values and Their Counts:")
for column in columns_with_nulls:
    print(f"{column}: {null_values[column]} null values")

Columns with Null Values and Their Counts:
workclass: 963 null values
occupation: 966 null values
native-country: 274 null values


### <span style="font-size: 16.38px;">Report: Variable Types for Columns with Null Values&nbsp;</span> 

### <span style="font-size: 16.38px;">Knowing what kinds of variables to look for while cleaning and preparing data is crucial since it will help you cope with null values. Because of this, columns containing null values were recognised and categorised according to the kind of variable that was being used.&nbsp;</span> 

### <span style="font-size: 16.38px;">Native-country, occupation, and workclass are the columns with null values. Upon examination, the aforementioned columns appear to belong to the following categories:</span>

### <span style="font-size: 16.38px;">Workclass: A Nominal Variable that categorises employment types in which a natural order does not exist.&nbsp;&nbsp;</span>  

### <span style="font-size: 16.38px;">Occupation: Nominal variable: this type of classification is used for work roles that lack a natural order.&nbsp;&nbsp;</span>  

### <span style="font-size: 16.38px;">Native-country: Nominal Variable - assigns a person's nation without any sort of natural hierarchy.&nbsp;</span> 

### <span style="font-size: 16.38px;">In conclusion, it is clear from the analysis above that every column in our dataset containing nominal variables has a null value. Sorting them according to this category helps clarify how we should handle missing data and implement appropriate solutions for this type of non-ordered categorical data.</span>

**<u>2.2 b)</u>**

As a result of the aforementioned analysis, it is evident that each column in our dataset that includes nominal variables has a null value. By classifying them, we may better understand how to deal with missing data and apply suitable fixes for this kind of non-ordered categorical data.

Python Code:

In [24]:
import pandas as pd

file_path = '/Users/maz/7CS033DataMining/Week2/Workshop/adult.csv'  
adult_data = pd.read_csv(file_path)

adult_data.dropna(inplace=True)

null_values_after_deletion = adult_data.isna().sum()
print(null_values_after_deletion)


age               0
workclass         0
fnlwgt            0
education         0
education-num     0
marital-status    0
occupation        0
relationship      0
race              0
sex               0
capital-gain      0
capital-loss      0
hours-per-week    0
native-country    0
income            0
dtype: int64


Implementation: Open pandas DataFrame and load the dataset from the specified file path. Verify that the file\_path variable refers to the location of your dataset.

Deleting Rows Where Null Values are Present: The dropna() method can be performed on the DataFrame itself to remove rows where null values are present. You can assign the result to a new variable because the inplace=True parameter alters the DataFrame in place. 

Verifying Removal of Nulls: Lastly, use isna() to count the number of nulls in each column of the updated DataFrame and display the result.sum(). Once null rows are removed, all columns will have 0 nulls, indicating that there are no missing data points in the dataset. 

Expected Output: 

Following the execution of this code, each cell in the output should show zeros, indicating that the dataset is now free of null values and that the programme has successfully removed any rows with missing data. In many types of data analysis and machine learning models, when working with null values becomes crucial for correct findings, it assures that the dataset has now been stripped off incomplete entries. 

Results: 

There are 0 null values given in every column, indicating that every row in the dataset has all of the missing data.

**<u>2.2 c)</u>**

Here is how this can be done: 

Here's how to go about doing it: Bring the dataset in: Let's start by reimporting the dataset using the right file path you provided. 

Finding categorical columns with null values: Based on our earlier study, we were able to determine that the categories workclass, occupation, and native country all contained null values. 

Find the mode and fill in the null values: We will first determine the mode for each of these columns before using it to fill in the null value. 

Findings: 

There are 0 null values given in every column, indicating that every row in the dataset has all of the missing data.  

### Step-by-Step Code:

In [25]:
import pandas as pd

file_path = '/Users/maz/7CS033DataMining/Week2/Workshop/adult.csv'

adult_data = pd.read_csv(file_path)

categorical_columns_with_nulls = ['workclass', 'occupation', 'native-country']

print("Initial Null Values Count:")
print(adult_data[categorical_columns_with_nulls].isna().sum())

for column in categorical_columns_with_nulls:
    mode_value = adult_data[column].mode()[0]  
    print(f"Mode for {column}: {mode_value}")  
    adult_data[column].fillna(mode_value, inplace=True)  

print("\nNull Values Count After Filling:")
print(adult_data[categorical_columns_with_nulls].isna().sum())

Initial Null Values Count:
workclass         963
occupation        966
native-country    274
dtype: int64
Mode for workclass: Private
Mode for occupation: Prof-specialty
Mode for native-country: United-States

Null Values Count After Filling:
workclass         0
occupation        0
native-country    0
dtype: int64


### <span style="font-size: 16.38px;">Explanation:&nbsp;</span> 

### <span style="font-size: 16.38px;">First, let's fill a pandas DataFrame with the dataset.&nbsp;</span> 

### <span style="font-size: 16.38px;">Null values in categorical columns: Workclass, occupation, and native-country were determined to be the categorical columns in need of null value treatment based on previous analysis.&nbsp;</span> 

### <span style="font-size: 16.38px;">Filling Null Values with Mode: We determine the mode, or most frequent value, for each categorical column containing nulls, and we use this mode to fill in the missing values. In order to maintain the variable's distribution, this is frequently done when handling missing data for categorical variables.&nbsp;</span> 

### <span style="font-size: 16.38px;">Verifying Changes: To ensure that the populating process was successful, we verify that there are still null values in these columns after the null values have been populated.&nbsp;</span> 

### <span style="font-size: 16.38px;">This is quite useful as it doesn't change the dataset's overall distribution, which is especially useful when working with categorical data. When the code above is executed, the given categorical columns will no longer contain null values; instead, the most common value found in each column will be filled in.</span>

## TASK 2.3. Select one of the columns which contains wrong data. Write a code to fill the wrong cells with an appropriate value. You have the freedom to determine what value you want to use to fill the wrong cells (2%). 

In TASK 2.3, we will select a column containing erroneous data and determine the appropriate approach to replace values that violate the law of consistency or contain errors. An excellent choice for the workclass column, based on our recent conversation, is the placeholder value "?", which we identified as signifying missing or erroneous data. 

Since using the mode is a standard procedure for addressing missing categorical data, let's assume that we would like to replace these placeholder values with the mode value of the data within the workclass column. This method makes the assumption that the mode, particularly in the lack of more evidence to infer a more precise fill value, is the most likely category for missing data.

Step-by-Step Code: 

- **Determine the Mode: In the "workclass" column, find the most frequently occurring value where the placeholder question mark is equivalent to "?".**
- **Replace Placeholders:** Substitute the discovered potential mode for the "?" values.
- **Verify the Replacement:** After the replacement, count the number of times "?" appears to ensure that it was successful.

Here's how you can do that:

In [26]:
import pandas as pd

file_path = '/Users/maz/7CS033DataMining/Week2/Workshop/adult.csv'

adult_data = pd.read_csv(file_path)

mode_value = adult_data.loc[adult_data['workclass'] != '?', 'workclass'].mode()[0]

adult_data['workclass'] = adult_data['workclass'].replace('?', mode_value)

remaining_placeholders = (adult_data['workclass'] == '?').sum()

print(f"Mode value used for replacement in 'workclass': {mode_value}")
print(f"Count of '?' placeholders after replacement in 'workclass': {remaining_placeholders}")

Mode value used for replacement in 'workclass': Private
Count of '?' placeholders after replacement in 'workclass': 0


### <span style="font-size: 16.38px;">Explanation:&nbsp;</span> 

### <span style="font-size: 16.38px;">Finding Mode: To make sure the value in the workclass column truly relates to a category, we first identify the value that appears the most frequently in the column by eliminating any "?" values.&nbsp;</span> 

### <span style="font-size: 16.38px;">Replacing Placeholders: In the next step, we enter this mode value in lieu of every "?" that appears in the workclass column. The replace() method is used to accomplish this.</span>

### <span style="font-size: 16.38px;">Verification: In the end, we count the number of "?" values that remain in the workclass column to determine the success of the replacement. If all of the replacements were successful, this number should be 0.&nbsp;</span> 

### <span style="font-size: 16.38px;">By replacing inaccurate data wherever it appears with a fair approximation based on the data's current distribution in the column, this offers a simple method of addressing faulty data.</span>

## TASK 2.4: Complete the following tasks: 

### a. Assume we want to predict the income column as the output. Use the correct method to encode this column (1%). 

### b. Select one ordinal column and encode it using the appropriate method (1%) . 

### c. Select one nominal column and encode it using the appropriate method (1%).

### d. After encoding the nominal column in the previous step, find a method to append the encoded data to the dataset (2%). 

### e. Normalise the numerical columns of the dataset (1%)

**<u>2.4 a)</u>**

If the income column is categorical—which is frequent for machine learning classification tasks—you'll probably want to encode it in order to forecast the income column as the output. In most cases, the Adult dataset's income column has two categories—"\>50K" and "\<=50K"—that indicate whether a person makes more, equal to, or less than $50,000 per year. For the majority of machine learning methods, it is necessary to encode this column into a numerical format.  

### Method for Encoding: Label Encoding

Label encoding is a popular technique for encoding binary categorical variables, in which a distinct number is assigned to each category. We can apply this strategy since the income column is binary, meaning it can have just two distinct values. You may accomplish this with pandas by directly mapping the categories to integers or, for a more automated method, by using LabelEncoder from sklearn.preprocessing.  

### Using pandas for Direct Mapping:

Here's how you can manually encode the income column using pandas by mapping the categories to integers:

In [27]:
income_encoded = adult_data['income'].map({'>50K': 1, '<=50K': 0})

adult_data['income_encoded'] = income_encoded

adult_data[['income', 'income_encoded']].head()

,income,income_encoded
0,<=50K,0.0
1,<=50K,0.0
2,<=50K,0.0
3,<=50K,0.0
4,<=50K,0.0


### Using LabelEncoder:

As an alternative, you can use LabelEncoder for a more automated method; this is particularly helpful if you want to use a library-based solution or are working with more than two categories:

In [28]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

adult_data['income_encoded'] = le.fit_transform(adult_data['income'])

adult_data[['income', 'income_encoded']].head()

,income,income_encoded
0,<=50K,0
1,<=50K,0
2,<=50K,0
3,<=50K,0
4,<=50K,0


<span style="font-size: 16px; white-space-collapse: preserve;">Using either approach, your DataFrame will get a new column called income_encoded, where the encoded values stand in for the income column's categories. Depending on your desire for explicit mapping over automatic encoding and the complexity of your dataset, you may need to decide between utilising LabelEncoder and manual mapping.</span>

**<u>2.4 b)</u>**

Education is an example of an ordinal column for the Adult dataset. From lower to higher levels of education, the achievements in education are ranked in this column. We can translate the education levels to integers that indicate their rank order in order to suitably capture this ordinal column.

<span style="font-family: -apple-system, BlinkMacSystemFont, sans-serif; color: var(--vscode-foreground);">Method for Encoding: Manual Mapping</span>  

Manual mapping is an appropriate technique since ordinal data reflects categories with a meaningful order. This makes it possible for us to rank the categories according to integers. The precise mapping is determined by the dataset's ranking of the education levels.  

Using the assumption that the dataset is arranged in a usual order from no schooling to advanced degrees, below is a potential mapping based on common education levels. The mapping may require modification in light of the real unique values in your education column:

In [29]:
education_mapping = {
    'Preschool': 0,
    '1st-4th': 1,
    '5th-6th': 2,
    '7th-8th': 3,
    '9th': 4,
    '10th': 5,
    '11th': 6,
    '12th': 7,
    'HS-grad': 8,
    'Some-college': 9,
    'Assoc-voc': 10,
    'Assoc-acdm': 11,
    'Bachelors': 12,
    'Masters': 13,
    'Prof-school': 14,
    'Doctorate': 15
}

adult_data['education_encoded'] = adult_data['education'].map(education_mapping)

adult_data[['education', 'education_encoded']].head()


,education,education_encoded
0,Bachelors,12
1,Bachelors,12
2,HS-grad,8
3,11th,6
4,Bachelors,12


With the help of this code snippet, the education column is manually encoded and saved as a new column called education\_encoded. There, each education level is represented by an integer based on its rank. The content and hierarchy of education levels in your dataset should be taken into consideration while modifying the mapping dictionary education\_mapping. 

Because manual mapping makes it possible to clearly identify the order of the categories—a necessary step in preserving the meaningful relationship between them—it is especially helpful for ordinal data.

**<u>2.4 c)</u>**

One-hot encoding is a suitable and often utilised technique for encoding a nominal column without suggesting any hierarchy among its categories. Models can handle each category as an independent feature without assuming any order by using this strategy, which converts each categorical value into a new binary column. 

Let's choose to use one-hot encoding and encode the nominal column workclass. To do this, we'll use the simple and effective get\_dummies() method in pandas, which creates one-hot encoded variables.  

### Step-by-Step Code:

In [31]:
import pandas as pd

file_path = '/Users/maz/7CS033DataMining/Week2/Workshop/adult.csv'
adult_data = pd.read_csv(file_path)

workclass_encoded = pd.get_dummies(adult_data['workclass'], prefix='workclass')

workclass_encoded.head()

,workclass_?,workclass_Federal-gov,workclass_Local-gov,workclass_Never-worked,workclass_Private,workclass_Self-emp-inc,workclass_Self-emp-not-inc,workclass_State-gov,workclass_Without-pay
0,False,False,False,False,False,False,False,True,False
1,False,False,False,False,False,False,True,False,False
2,False,False,False,False,True,False,False,False,False
3,False,False,False,False,True,False,False,False,False
4,False,False,False,False,True,False,False,False,False


With the help of this code snippet, a new DataFrame called workclass\_encoded will be created, with each workclass category represented as a distinct column that is prefixed with workclass\_. For every row, the column that corresponds to its category will have a 1; all other columns will have 0s. 

The rationale behind pd.get\_dummies() The workclass column in the adult\_data table is the target of this function. In order to identify the source column (workclass) after encoding, we apply a prefix to the new columns using the prefix option. 

The workclass\_encoded as a result For every category in the original workclass column, DataFrame includes binary columns. This successfully transforms the nominal data into a format that is appropriate for machine learning algorithms without suggesting any kind of ordinal relationship.  

When handling nominal data for machine learning preprocessing, this approach is very helpful because it guarantees that the model can distinguish between different categories without making any unintentional ordinal assumptions.

**<u>2.4 d)</u>**

Once a nominal column, such as workclass, has been encoded using one-hot encoding, you may use the pandas library to append the encoded data back to the original dataset. When preparing a dataset for machine learning models, which need numerical input, this is frequently done. Concatenating the original DataFrame and the one-hot encoded DataFrame is the procedure. 

The one-hot encoded workclass data can be appended to the original adult\_data DataFrame in the following way:

In [32]:
import pandas as pd

file_path = '/Users/maz/7CS033DataMining/Week2/Workshop/adult.csv'  

adult_data = pd.read_csv(file_path)

workclass_encoded = pd.get_dummies(adult_data['workclass'], prefix='workclass')

adult_data_encoded = pd.concat([adult_data, workclass_encoded], axis=1)

adult_data_encoded.head()

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,...,income,workclass_?,workclass_Federal-gov,workclass_Local-gov,workclass_Never-worked,workclass_Private,workclass_Self-emp-inc,workclass_Self-emp-not-inc,workclass_State-gov,workclass_Without-pay
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,...,<=50K,False,False,False,False,False,False,False,True,False
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,...,<=50K,False,False,False,False,False,False,True,False,False
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,...,<=50K,False,False,False,False,True,False,False,False,False
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,...,<=50K,False,False,False,False,True,False,False,False,False
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,...,<=50K,False,False,False,False,True,False,False,False,False


### Explanation:

- Concatenation: The adult\_data DataFrame and the workclass\_encoded DataFrame are joined using pd.concat(). By concatenating the additional one-hot encoded columns to the right side of the original DataFrame, the axis=1 option essentially adds the new columns to the original DataFrame. 
- Removing the Original Column: To reduce redundancy in the dataset, you may decide to remove the original workclass column after encoding and appending the new columns. Whether you need to keep the original column for reference or additional analysis will determine whether you should do this optionally. 
- Verification: You can confirm that the one-hot encoded columns have been successfully added to the dataset by using head() to display the first few rows of the modified DataFrame.

You can prepare your original dataset for future data processing or machine learning activities by integrating the encoded data using these procedures. Using this method guarantees that your dataset makes use of the data found in nominal columns in a way that makes it compatible with numerical techniques.

**<u>2.4 e)</u>**

There are several methods you can use to normalise the dataset's numerical columns. Using Min-Max normalisation, which scales the data in each column to a given range (often \[0, 1\]), is a popular method. With the help of the MinMaxScaler from sklearn.preprocessing, this is readily accomplished. 

In machine learning, normalisation is especially helpful in preventing bias resulting from feature scale by ensuring that numerical columns contribute equally to the model's training process. 

The following steps will help you normalise the dataset's numerical columns: 

Step 1, Determine the Numerical Columns. Determine which of your dataset's columns are numerical first. Number columns such as age, fnlwgt, education-num, capital-gain, capital-loss, and hours-per-week are common examples, assuming the Adult dataset. 

Step 2: Use Normalisation (Min-Max)

In [33]:
from sklearn.preprocessing import MinMaxScaler
import pandas as pd

file_path = '/Users/maz/7CS033DataMining/Week2/Workshop/adult.csv'  

adult_data = pd.read_csv(file_path)

numerical_columns = ['age', 'fnlwgt', 'education-num', 'capital-gain', 'capital-loss', 'hours-per-week']

scaler = MinMaxScaler()

adult_data[numerical_columns] = scaler.fit_transform(adult_data[numerical_columns])

adult_data.head()

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
0,0.301370,State-gov,0.044131,Bachelors,0.800000,Never-married,Adm-clerical,Not-in-family,White,Male,0.02174,0.0,0.397959,United-States,<=50K
1,0.452055,Self-emp-not-inc,0.048052,Bachelors,0.800000,Married-civ-spouse,Exec-managerial,Husband,White,Male,0.00000,0.0,0.122449,United-States,<=50K
2,0.287671,Private,0.137581,HS-grad,0.533333,Divorced,Handlers-cleaners,Not-in-family,White,Male,0.00000,0.0,0.397959,United-States,<=50K
3,0.493151,Private,0.150486,11th,0.400000,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0.00000,0.0,0.397959,United-States,<=50K
4,0.150685,Private,0.220635,Bachelors,0.800000,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0.00000,0.0,0.397959,Cuba,<=50K


### <span style="font-size: 16.38px;">Reasoning: Recognising Numerical Columns Normalisation should only be applied to the numerical columns because scaling categorical data in this manner would be illogical and might generate mistakes in the data.&nbsp;</span> 

### <span style="font-size: 16.38px;">MinMaxScaler: This sklearn utility.Preprocessing is the act of transforming and scaling the data so that each attribute in the input data is scaled to fall inside one of two ranges: [0, 1] or an otherwise specified range.&nbsp;</span> 

### <span style="font-size: 16.38px;">Fitting and Transforming: To scale the data to the designated range, the fit_transform() method calculates the minimum and maximum values of the data to be changed. You may guarantee that the values in your dataset are confined inside a particular range by normalising the numerical columns. This will enable more stable and rapid convergence while training machine learning models.</span>